ts-types/ensure.d.ts

```diff
@@ -0,0 +1,16 @@
+export type EnsureFunction = (...args: any[]) => any;
+export interface EnsureOptions {
+	name?: string;
+	isOptional?: boolean;
+	default?: any;
+	errorMessage?: string;
+	errorCode?: number;
+	Error?: ErrorConstructor;
+}
+
+type ValidationDatum = [argumentName: string, inputValue: any, ensureFunction: EnsureFunction, options?: object];
+type ValidationDatumList = ValidationDatum[];
+
+declare function ensure<T>(...args: [...ValidationDatumList, EnsureOptions]): T;
+declare function ensure<T>(...args: [...ValidationDatumList]): T;
```

Comment:
Now, this should work as expected but, it will work fine only on ts version 4 or higher.

Reply:
It's fine. I think it's ok to limit it to TS>= 4. WDYT? What is the percentage of usage below that version?

Reply:
Other question: `ensure` takes array and returns array, but it's a very different array, Input is array of arrays, and output is array of resolved values. Will `<T>` work in such case?

Reply:
Other question: ensure takes array and returns array, but it's a very different array, Input is array of arrays, and output is array of resolved values. Will <T> work in such case?

Reply:
@medikoo yes you can have a type as we have in `type ValidationDatum` and use a similar strategy.

```
type StrNumTuple = [string, number];

function test<T>(a: any, b: any): T;
function test(a: any, b: any) {
   const c = [a.toString(), +b];
   return c;
}

var c = test<StrNumTuple>('o', 2);
var [d, e] = c;
console.log(d);
console.log(e);
```

Reply:
@borracciaBlu I'm not sure if it answers my question.

What I meant, is that `ensure` takes array in format as `[[string, any, options?], ...[string, any, options?], options?]` while returns the array in format `[any, any, any]`

So is `<T>(): T` applicable to such case? I assume that it's ok, as it'll just validate that both _input_ and _output_ are arrays, but won't dive into internal structure of them, right?

Reply:
about stats `TS >= 4` is ~ 80%  and i would guess it's growing...

```
v 0: 0.003%, 
v1: 0.112%, 
v2: 2.106% => 559,038, 
v3: 20.04% => 5,317,892, 
v4: 78.705%
```

Reply:
Hi @medikoo yes it is.

**Input**
```
[[string, any, options?], ...[string, any, options?], options?]

this is because 

type ValidationDatum = [argumentName: string, inputValue: any, ensureFunction: EnsureFunction, options?: object];
type ValidationDatumList = ValidationDatum[];
(...args: [...ValidationDatumList])
```

**Output**
for `<T>(): T`:

```
you can define a type 
eg: 
type StrNumTuple = [string, number];

and then 

let tuple: StrNumTuple = ensure<StrNumTuple>(["foo", 12.323, ensureString], ["bar", 10, ensureNumber]);
```